Source/WebCore/platform/ScrollSnapAnimatorState.cpp

```diff
@@ -95,49 +95,81 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
-}
-
-bool ScrollSnapAnimatorState::preserveCurrentTargetForAxis(ScrollEventAxis axis)
-{
-    auto snapID = activeSnapIDForAxis(axis);
-    auto snapOffsets = snapOffsetsForAxis(axis);
     
-    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [snapID](SnapOffset<LayoutUnit> p) -> bool {
-        return p.snapTargetID == *snapID;
+    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [boxID](SnapOffset<LayoutUnit> p) -> bool {
+        return p.snapTargetID == boxID;
     });
     if (found == snapOffsets.end()) {
-        setActiveSnapIndexIDForAxis(axis, std::nullopt);
+        setActiveSnapIndexForAxis(axis, std::nullopt);
         return false;
     }
     
     setActiveSnapIndexForAxis(axis, std::distance(snapOffsets.begin(), found));
     return true;
 }
 
-bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const ScrollExtents& scrollExtents, float pageScale)
+Vector<SnapOffset<LayoutUnit>> ScrollSnapAnimatorState::currentlySnappedOffsetsForAxis(ScrollEventAxis axis) const
 {
-    // Check if we need to set the active target to a focused element
-    setFocusedElementForAxis(ScrollEventAxis::Vertical);
-    setFocusedElementForAxis(ScrollEventAxis::Horizontal);
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsets;
+    auto snapOffsets = snapOffsetsForAxis(axis);
+    auto activeIndex = activeSnapIndexForAxis(axis);
+    
+    if (activeIndex && *activeIndex < snapOffsets.size())
+        currentlySnappedOffsets.append(snapOffsets[*activeIndex]);
+    return currentlySnappedOffsets;
+}
+
+static bool isSnappedToMultipleBoxes(const Vector<SnapOffset<LayoutUnit>>& verticalOffsets, const Vector<SnapOffset<LayoutUnit>>& horizontalOffsets)
+{
+    HashSet<ElementIdentifier> boxIDs;
+    
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsets;
+    currentlySnappedOffsets.appendVector(verticalOffsets);
+    currentlySnappedOffsets.appendVector(horizontalOffsets);
     
+    bool hasMultipleSnapAreas = false;
+    for (auto offset : currentlySnappedOffsets) {
+        boxIDs.add(offset.snapTargetID);
+        hasMultipleSnapAreas |= offset.snapAreaIndices.size() > 1;
+    }
+    return boxIDs.size() > 1 || hasMultipleSnapAreas;
+}
+
+ElementIdentifier ScrollSnapAnimatorState::chooseBoxToResnapTo(const Vector<SnapOffset<LayoutUnit>>& verticalOffsets, const Vector<SnapOffset<LayoutUnit>>& horizontalOffsets)
+{
+    HashSet<ElementIdentifier> boxIDs;
+    
+    Vector<SnapOffset<LayoutUnit>> currentlySnappedOffsets;
+    currentlySnappedOffsets.appendVector(verticalOffsets);
+    currentlySnappedOffsets.appendVector(horizontalOffsets);
+    
+    for (auto offset : verticalOffsets) {
+        boxIDs.add(offset.snapTargetID);
+        for (auto i : offset.snapAreaIndices)
+            boxIDs.add(m_snapOffsetsInfo.snapAreasIDs[i]);
+    }
+    
+    Vector<SnapOffset<LayoutUnit>> possibleOffsets;
+    possibleOffsets.appendVector(snapOffsetsForAxis(ScrollEventAxis::Horizontal));
```

Comment:
I was confused by the relationship between the arguments (e.g. `horizontalOffsets `) and what `snapOffsetsForAxis(ScrollEventAxis::Horizontal)` gives you. I think the argument is "current" and "snapOffsetsForAxis()" is computing the new state? It would be clearer to call `snapOffsetsForAxis()` in the caller, and explicitly pass the new ones. This function can access `m_currentlySnappedBoxesX` and  `m_currentlySnappedBoxesY` for the current state, right?

Reply:
You could even make a helper function that just returns the `HashSet<ElementIdentifier>` of currently snapped boxIDs.

Reply:
Ah I see yes the idea is comparing the currently snapped boxes (`m_currentlySnappedBoxesX` and `m_currentlySnappedBoxesY`) and the newly snapped boxes (snapOffsetsForAxis). You are right that it makes more sense to pass the newly snapped boxes and access the member variables from this function. Also added the suggested helper function `currentlySnappedBoxes()`.